src/Storages/StorageMergeTree.cpp

```diff
@@ -1428,72 +1409,288 @@ MergeTreeDataPartPtr StorageMergeTree::outdatePart(MergeTreeTransaction * txn, c
     }
 
-    dropPartsImpl(std::move(parts_to_remove), detach);
+    return partition_infos;
 }
 
-void StorageMergeTree::dropPartsImpl(DataPartsVector && parts_to_remove, bool detach)
+MergeTreeData::MutableDataPartsVector StorageMergeTree::makeCoveringEmptyTmpParts(const DataPartsVector & parts, const MergeTreeTransactionPtr & txn)
 {
-    auto metadata_snapshot = getInMemoryMetadataPtr();
+    MutableDataPartsVector new_parts;
 
-    if (detach)
+    /// Parts could have gaps in continuous block numeration of parts due to concurrent insertions.
+    /// This gaps must not be covered with empty parts otherwise parts from concurrent insertions miss.
+    /// Here the parts are split on ranges without block gaps.
+    RangesWithContinuousBlocks continuous_ranges = groupByRangesWithContinuousBlocks(parts);
+
+    for (auto& range: continuous_ranges)
     {
-        /// If DETACH clone parts to detached/ directory
-        /// NOTE: no race with background cleanup until we hold pointers to parts
-        for (const auto & part : parts_to_remove)
+        /// Inside each range new empty part could cover parts from min_block to max_block with respect of partitions
+        auto partitions = collectPartitionInfos(range);
+        for (const auto & it: partitions)
         {
-            LOG_INFO(log, "Detaching {}", part->data_part_storage->getPartDirectory());
-            part->makeCloneInDetached("", metadata_snapshot);
+            const auto & partition_id = it.first;
+            const auto & info = it.second;
+            auto part_info = MergeTreePartInfo(partition_id, info.min_block, info.max_block, info.max_level+1, info.max_mutation);
+
+            auto data_part = createEmptyPart(part_info, info.partition, part_info.getPartName(), txn);
+            new_parts.push_back(data_part);
         }
     }
 
-    if (deduplication_log)
+    auto get_part_names = [] (auto & parts_) -> Strings
     {
-        for (const auto & part : parts_to_remove)
+        Strings part_names;
+        for (const auto & p : parts_)
+            part_names.push_back(p->getNameWithState());
+        return part_names;
+    };
+
+    LOG_TEST(log, "Made {} empty parts in order to cover {} parts. Empty parts: {}, covered parts: {}",
+             new_parts.size(), parts.size(),
+             fmt::join(get_part_names(new_parts), ", "), fmt::join(get_part_names(parts), ", "));
+
+    return new_parts;
+}
+
+void StorageMergeTree::coverPartsWithEmptyParts(const DataPartsVector & old_parts, const MutableDataPartsVector & new_parts, Transaction & transaction)
+{
+    auto part_lock = lockParts();
+
+    /// Check that all old parts are Active still
+    {
+        DataPartsVector changed_parts;
+
+        for (const auto & part: old_parts)
+            if (part->getState() != DataPartState::Active)
+                changed_parts.push_back(part);
+
+        if (!changed_parts.empty())
```

Comment:
I forgot, why must old part be Active? This check is supposed to protect from concurrent TRUNCATE, but at the first glance there's nothing to protect from: if some new covering part is already created in another thread, then the corresponding `new_part` will be just ignored with "Tried to commit obsolete part" message on attempt to commit it.

Reply:
I checked this and the error message was unclear.
In case when some query drop part, that query create new one. 

    $CLICKHOUSE_CLIENT -q "insert into $table values (1)" # inserts all_1_1_0
    $CLICKHOUSE_CLIENT -q "insert into $table values (2)" # inserts all_2_2_0
    $CLICKHOUSE_CLIENT -q "insert into $table values (3)" # inserts all_3_3_0

    tx 11 "begin transaction"
    tx 22                         "begin transaction"
    tx 22                         "alter table tt drop part 'all_2_2_0'" # create all_2_2_1
    tx 11 "truncate table tt" | grep -Eo "PART_IS_TEMPORARILY_LOCKED" | uniq # try to create all_1_3_1
    tx 11 "commit" | grep -Eo "INVALID_TRANSACTION" | uniq
    tx 22                         "commit"

all_1_3_1 intersects with all_2_2_1. Intersections is forbidden, exception is thrown. User gets the incomprehensible error message. 
We have to either check that the operation is going to commit without collision or rework that protection inside commit.

Reply:
> As a result there is exception "Tried to commit obsolete part". 

But it's not an exception, it's just a warning and we can suppress it for this specific case